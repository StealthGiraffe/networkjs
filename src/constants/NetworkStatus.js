export const NetworkStatus = {
    ONLINE: 'online',
    OFFLINE: 'offline',
    STABLE: 'stable',
    UNSTABLE: 'unstable',
}

export const NetworkStatuses = Object.values(NetworkStatus)
