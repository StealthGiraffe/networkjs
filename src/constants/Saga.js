export const Saga = {
    NETWORK: 'network',
    STABILITY: 'stability',
}

export const Sagas = Object.values(Saga)
