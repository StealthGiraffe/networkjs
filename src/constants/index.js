export { NetworkSpeed } from './NetworkSpeed'
export { NetworkStatus, NetworkStatuses } from './NetworkStatus'
export { Saga, Sagas } from './Saga'
