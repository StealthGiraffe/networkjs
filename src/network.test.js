import { NetworkStatus, NetworkStatuses, Monitor, Monitors, ErrorMessage } from './constants'
import NetworkMonitor from './monitors/network'
import ServiceMonitor from './monitors/service'
import StabilityMonitor from './monitors/stability'
import EventEmitter from './events/EventEmitter'
import Network from './index'

jest.mock('./events/EventEmitter')
jest.mock('./monitors/service')
jest.mock('./monitors/stability')
jest.mock('./monitors/network')

describe('Network', () => {
    let monitor
    const emitter = new EventEmitter()
    const networkMonitor = new NetworkMonitor() 
    const stabilityMonitor = new StabilityMonitor() 
    const serviceMonitor = new ServiceMonitor() 
    
    beforeEach(() => {
        window.PerformanceObserver = 'arbitraryTest'
        monitor = new Network()
    })

    afterEach(() => {
        jest.clearAllMocks()
    })

    describe('constructor', () => {
        it('initializes with the correct props', () => {
            //TODO: refactor this to only check certain properties
            expect(monitor).toMatchSnapshot()
        })
    })

    describe('on', () => {
        it('throws error on string not in NetworkStatuses', () => {
            expect(() => {
                monitor.on('INVALID', jest.fn)
            }).toThrow(`${ErrorMessage.INVALID_EVENT}`)
        })
   
        it.each`
            networkStatus             
            ${NetworkStatus.ONLINE}      
            ${NetworkStatus.OFFLINE}      
            ${NetworkStatus.STABLE}      
            ${NetworkStatus.UNSTABLE}      
            ${NetworkStatus.DEGRADED}      
            ${NetworkStatus.RESOLVED}      
        `('adds an EventListener to the $networkStatus monitor eventEmitter', ({networkStatus}) => {
            monitor.on(networkStatus, jest.fn)
            expect(monitor.eventEmitter.addEventListener).toHaveBeenCalledTimes(
                1,
            )
        })
    })

    describe('all', () => {
        it('Adds an EventListener for each NetworkStatus', () => {
            monitor.all(jest.fn)
            expect(monitor.eventEmitter.addEventListener).toHaveBeenCalledTimes(
                NetworkStatuses.length,
            )
        })
    })

    describe('serviceError', () => {
        it('Calls handleError from the service monitor', () => {
            monitor.serviceError('/api/resource/1', 502)
            expect(
                monitor.monitors[Monitor.SERVICE].handleError,
            ).toHaveBeenCalled()
        })
    })
    
    describe('pause', () => {
        it('calls pause on the given monitor', () => {
            monitor.pause(Monitor.SERVICE)
            expect(monitor.monitors[Monitor.SERVICE].pause).toHaveBeenCalledTimes(1)
        })
        
        it('calls pause on all monitors', () => {
            monitor.pause()
            
            Monitors.forEach( m => {
                expect(monitor.monitors[m].pause).toHaveBeenCalledTimes(1)
            })
        })
    })

    describe('resume', () => {
        it('calls resume on the given monitor', () => {
            monitor.resume(Monitor.SERVICE)
            expect(monitor.monitors[Monitor.SERVICE].resume).toHaveBeenCalledTimes(1)
        })

        it('calls resume on all monitors', () => {
            monitor.resume()

            Monitors.forEach(m => {
                expect(monitor.monitors[m].resume).toHaveBeenCalledTimes(1)
            })
        })
    })
})
