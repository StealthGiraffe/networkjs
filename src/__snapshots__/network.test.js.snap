// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network constructor initializes with the correct props 1`] = `
Network {
  "eventEmitter": EventEmitter {
    "addEventListener": [MockFunction],
    "dispatchEvent": [MockFunction],
    "registerEvent": [MockFunction] {
      "calls": Array [
        Array [
          "online",
        ],
        Array [
          "offline",
        ],
        Array [
          "stable",
        ],
        Array [
          "unstable",
        ],
        Array [
          "degraded",
        ],
        Array [
          "resolved",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  "monitors": Object {
    "network": NetworkMonitor {
      "initialize": [MockFunction],
      "pause": [MockFunction],
      "resume": [MockFunction],
    },
    "service": ServiceMonitor {
      "decrement": [MockFunction],
      "getServiceForUrl": [MockFunction],
      "handleError": [MockFunction],
      "increment": [MockFunction],
      "initialize": [MockFunction],
      "pause": [MockFunction],
      "resume": [MockFunction],
      "waitAndDecrement": [MockFunction],
    },
    "stability": StabilityMonitor {
      "_addEntry": [MockFunction],
      "_adjustEntryWeights": [MockFunction],
      "_emitStabilityChanges": [MockFunction],
      "_removeOverflowEntry": [MockFunction],
      "initialize": [MockFunction],
      "pause": [MockFunction],
      "resume": [MockFunction],
      "run": [MockFunction],
    },
  },
}
`;
